### PR TITLE
Builder equality check

### PR DESCRIPTION
**Pull request by liamcain**

### CODE REVIEW

File: src/BaseBuilder.js [hunk 1]
```diff
@@ -4,6 +4,11 @@ import tcomb from 'tcomb-validation';
 import { validation } from './combinators';
 import * as validators from './validators';
 
+function keyIn(...keys) {
```
Comment (dblarons) [inline src/BaseBuilder.js:7]: Can you put this in the utils file? https://github.com/cadre/tcomb-builder/blob/master/src/utils.js

File: src/__tests__/BaseBuilder.spec.js [hunk 2]
```diff
@@ -482,4 +482,58 @@ describe('BaseBuilder', () => {
...
+
+  describe('isEqual()', () => {
+    it('returns true with passed the same instance', () => {
+      const builder = new BaseBuilder();
+      expect(builder.isEqual(builder)).to.be.true;
+    });
+
+    it('can compare equivalent builders', () => {
+      const builder = new BaseBuilder();
+      expect(builder.isEqual(builder)).to.be.true;
```
Comment (dblarons) [inline src/__tests__/BaseBuilder.spec.js:494]: Remove this test–has same body as the one above it and same description as one below it

File: src/__tests__/BaseBuilder.spec.js [hunk 3]
```diff
@@ -482,4 +482,58 @@ describe('BaseBuilder', () => {
...
+
+    it('can compare inequivalent builders', () => {
+      const error = () => 'Error';
+      const error2 = () => null;
+      const builder1 = new BaseBuilder()
+        .setValidationErrorMessageFn(error);
+      const builder2 = new BaseBuilder()
+        .setValidationErrorMessageFn(error2);
+
+      expect(builder1.isEqual(builder2)).to.be.false;
```
Comment (dblarons) [inline src/__tests__/BaseBuilder.spec.js:536]: What happens when:

```
const error = () => 'Error';
const error2 = () => 'Error';
```

?

File: src/primitives/__tests__/UnionBuilder.spec.js [hunk 4]
```diff
@@ -77,4 +77,32 @@ describe('UnionBuilder', () => {
...
+      const otherUsa = StructBuilder
+        .setField('country', TextBuilder)
+        .setField('state', TextBuilder);
+      const otherCountriesBuilder = UnionBuilder
+        .setUnion([otherInternational, otherUsa])
+        .setDispatch(value => (
+          value.country === 'usa' ? otherUsa : otherInternational
+        ));
+
+      expect(countriesBuilder.isEqual(otherCountriesBuilder)).to.be.true;
```
Comment (dblarons) [inline src/primitives/__tests__/UnionBuilder.spec.js:105]: Can you add a test case where two unionbuilders are not equal? Perhaps where their dispatch is not the same.

**Review by dblarons — COMMENTED** [review]
In general, this solves our use-case. The main problem, as I'm sure you're aware, is that without being able to check for equality in callback functions (e.g. `_dispatch`), the builders aren't certain to be equal. As a user of tcomb-builder, I would expect `isEqual` to check for full equality.

An example:

```
const countriesBuilder1 = UnionBuilder
  .setUnion([international, usa])
  .setDispatch(value => (
    value.country === 'usa' ? usa : international
));

const countriesBuilder2 = UnionBuilder
  .setUnion([international, usa])
  .setDispatch(value => (
    value.country !== 'usa' ? international, usa
));

countriesBuilder1.isEqual(countriesBuilder2); => returns true, even though they are not equal
```

In tcomb, types have names. I recommend that we do something similar for error and dispatch functions, as well as enforcing names for types in the builder, so that those names can be checked for equality. It's not a perfect solution, but it brings the `isEqual` function much closer to what its name implies.

That said, interested to hear thoughts and other proposed solutions.

File: src/__tests__/BaseBuilder.spec.js (hunk 3 above)
Comment (liamcain) [reply src/__tests__/BaseBuilder.spec.js:536]: Those aren't equal. error !== error2. Which is fine since all our errors are stored in const files.

**Comment by dblarons** [comment]
Spoke in person – @BoundInCode has a fix for the specific problem that prompted this PR at the component level. Don't urgently need this, so we'll take a bit more time to think about the best solution.

**Pull request closed without merging**